### PR TITLE
Refactor of confd and add k8s backend

### PR DESCRIPTION
**Pull request by heschlie**
As we are planning to migrate off of confd in the long run I refactored
the code to be a bit more like an actual go repo and switched to Glide
for dependency management.

I have also added a very naive k8s backend that accounts for everything
we currently store in k8s, we will need to add more when we populate the
rest.

The k8s backend uses the "defualt" client setup (it will autodetect settings when run in a cluster) and can be overridden with the `-kubeconfig` flag if you want to use it outside a cluster.

- Moved the contents of `src/github.com/kelseyhightower/confd/*` into
  the top level directory
- Migrated to Glide for dependencies
- Added k8s backend to pull needed "keys" from k8s
- `k8s.Watch()` simply refreshes every 10 seconds, should probably just be
  omitted in favor of `-interval`

### CODE REVIEW

**Comment by caseydavenport** [comment]
@heschlie would it be possible to open a version of this PR without all the vendor changes? Pretty hard to review as it is :)

**Comment by heschlie** [comment]
@caseydavenport You mean you don't want to scroll through ~4800 files?  hmm I guess that seems reasonable, will post updated PR

**Comment by heschlie** [comment]
@caseydavenport @robbrockbank I've reset the vendor directory should be less junk to sift through

File: backends/k8s/client.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+					}
+				}
+			}
+		}
+	}
+	fmt.Printf("%v", kvps)
+	return kvps, nil
+}
+
+func (c *Client) WatchPrefix(prefix string, waitIndex uint64, stopChan chan bool, keys []string) (uint64, error) {
```
Comment (caseydavenport) [inline backends/k8s/client.go]: After discussing with @robbrockbank I think we probably do want to implement the watch stuff here.

File: backends/k8s/client.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+	K8sAPIEndpoint string
+	K8sKeyFile     string
+	K8sCertFile    string
+	K8sCAFile      string
+	K8sAPIToken    string
+}
+
+func NewK8sClient(kubeconfig string) (*Client, error) {
+
+	kc := KubeConfig {
```
Comment (heschlie) [inline backends/k8s/client.go]: Good point, I had removed all the CLI flags with host, port etc... This serves no point anymore.

File: backends/k8s/client.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+			host := m[len(m)-1]
+			node, err := c.clientSet.Nodes().Get(host, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+
+			var prefix = fmt.Sprintf("/calico/bgp/v1/host/%s/", node.Name)
+			for _, ip := range node.Status.Addresses {
+				if ip.Type == kapiv1.NodeInternalIP {
+					kvps[prefix + "ip_addr_v4"] = ip.Address
```
Comment (heschlie) [inline backends/k8s/client.go]: We indeed will want those later, but right now the templates check for those first, and take an action if they exist, if they do not they default to nothing or the global values.

Also, do we want to go ahead and get the stuff in for IPv6 as well?  I figured k8s didn't support it yet so we were going to skip it, but I can get the `bird6.cfg.mesh` setup as well.

File: backends/k8s/client.go [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+			// not exist
+			if err != nil {
+				if !kerrors.IsNotFound(err) {
+					return nil, err
+				}
+			}
+
+			for _, pool := range tprs.Items {
+				// We need to parse the JSON value and determine if it is a V4 CIDR
+				// This is ugly, might be better have this included in ippool metadata?
```
Comment (heschlie) [inline backends/k8s/client.go]: Much cleaner, thanks!

File: backends/k8s/client.go (hunk 3 above)
Comment (caseydavenport) [inline backends/k8s/client.go]: Nope, let's skip all the ones I listed as optional up there, and not worry about IPv6 yet since k8s doesn't fully support it.

My main concern is about `/host/%s/network_v4`, which we'll need to populate to make the `cross-subnet` IPIP type work.

We've got a couple of options:
- Implement the `Apply()` action on the Node resource in Kubernetes and write this info as an annotation.
- Just hard code this to a /24 somehow (scary).

Let's chat about what this means.

**Comment by heschlie** [comment]
@robbrockbank ready for your review now

File: backends/k8s/client.go [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+		switch key {
+		case global:
+			// Default to "info" until this makes it into k8s
+			kvps[globalLogging] = "info"
+			// Default to 64512
+			kvps[globalASN] = "64512"
+			// Default to False until peering info is available in k8s
+			kvps[globalNodeMesh] = `{"enabled": true}`
+		case globalNodeMesh:
+			// This is needed as there are calls to just 'global' and 'global/loglevel'
```
Comment (robbrockbank) [inline backends/k8s/client.go]: Is this comment correct?  I don't see a separate branch for global/loglevel

Comment (heschlie) [inline backends/k8s/client.go]: We have two templates that need it, [bird.toml.toml](https://github.com/projectcalico/calicoctl/blob/master/calico_node/filesystem/etc/calico/confd/conf.d/bird.toml.toml) and [bird6.toml.toml](https://github.com/projectcalico/calicoctl/blob/master/calico_node/filesystem/etc/calico/confd/conf.d/bird6.toml.toml).  They generate the bird.toml and bird6.toml files used for the bird.cfg and bird6.cfg.

Comment (robbrockbank) [inline backends/k8s/client.go]: Still not sure I follow - what do you mean by "calls to" - you mean it explicitly requests keys "global" and "global/loglevel"?  I guess I don't understand why this branch is for globalNodeMesh but the comment talks about global and global/loglevel.

File: backends/k8s/client.go [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+
+    	return kubeClient, nil
+}
+
+// Take the etcd like keys and route it to the appropriate k8s API endpoint
+func (c *Client) GetValues(keys []string) (map[string]string, error) {
+	var kvps = make(map[string]string)
+	for _, key := range keys {
+		log.Debug(fmt.Sprintf("Getting key %s", key))
+		if m := singleNode.FindStringSubmatch(key); m != nil {
```
Comment (robbrockbank) [inline backends/k8s/client.go]: I assume you only want to invoke this processing once per set of keys - at the moment it looks as though multiple keys could match and it'll be invoked multiple times.

File: config.go [hunk 7]
```diff
@@ -62,6 +63,7 @@ type Config struct {
 	Table        string   `toml:"table"`
 	LogLevel     string   `toml:"log-level"`
 	Watch        bool     `toml:"watch"`
+	KubeConfig   string   `toml:"kubeconfig"`
```
Comment (robbrockbank) [inline config.go]: Do we also need to support handling API endpoint/key/cert files being passed in explicitly or will we handle that in the calico/node by writing out a "local" kubeconfig file?

File: backends/k8s/client.go (hunk 5 above)
Comment (heschlie) [inline backends/k8s/client.go]: Ah sorry no I misread what you said, the comment is incorrect, that should read global/nodemesh not loglevel as you are correct I don't believe there are any direct calls to loglevel

File: backends/k8s/client.go (hunk 6 above)
Comment (heschlie) [inline backends/k8s/client.go]: Right, it definitely could if there are subdirectories in the keys... will think on this, we'll still need to parse each key that comes in but maybe track if we've already parse a particular host that run.

File: config.go (hunk 7 above)
Comment (heschlie) [inline config.go]: No the Go client for k8s is pretty clever, it will detect if it is running in a cluster and grab the appropriate info from there for you.  If you want to run this from outside the cluster you would need a kubeconfig to point to.

Comment (robbrockbank) [inline config.go]: I mean that in our API I believe you have the option of specifying either a cube config file OR the API endpoint/key/cert files.  So just wondered if we needed that here too.

**Comment by heschlie** [comment]
Some minor changes, still working on watch calls

File: backends/k8s/client.go (hunk 6 above)
Comment (heschlie) [inline backends/k8s/client.go]: So looking at our keys that are getting queried I don't think it is an issue, I did decide to omit the final optional `/` in the regex, and put an end anchor there, so it should suffice

**Comment by heschlie** [comment]
I've poked more at the watch calls here, I need to do more testing to see how well it is behaving but it seems to be working on my initial tests.

@caseydavenport I don't think the k8s go client exposes a way to watch a single node, does that sound correct to you or am I missing some way to tell either `ListOptions{}` or `GetOptions{}` to start a watch on a single node?

File: config.go (hunk 7 above)
Comment (caseydavenport) [inline config.go]: Ah, right.  Yeah, we don't need that here.

File: resource/template/resource.go [hunk 8]
```diff
@@ -81,6 +81,7 @@ func NewTemplateResource(path string, config Config) (*TemplateResource, error)
 		return nil, ErrEmptySrc
 	}
 	tr.Src = filepath.Join(config.TemplateDir, tr.Src)
+	log.Debug(fmt.Sprintf("Template Resource:\n%v", tr))
```
Comment (caseydavenport) [inline resource/template/resource.go]: Can we get rid of this guy?

File: test [hunk 9]
```diff
@@ -1,4 +1,4 @@
 #!/usr/bin/env bash
 
 echo "Running tests..."
-gb test
+go test
```
Comment (caseydavenport) [inline test]: Was this a typo? What's up with this change?

Comment (heschlie) [inline test]: This was using [gb](https://getgb.io/) for its vendoring and tests, I migrated it to Glide as I was running into problems getting gb working and to compile.  If we would prefer we can revert back to gb and revert these changes.

Comment (caseydavenport) [inline test]: OK. That's fine.

**Comment by caseydavenport** [comment]
Last few comments then LGTM.

**Comment by heschlie** [comment]
@caseydavenport Can I go ahead and include the vendor directory removal on the rebase?

**Comment by caseydavenport** [comment]
@heschlie go for it.

**Comment by heschlie** [comment]
@caseydavenport I removed the loop in the watch function as it wasn't needed, I also found that we needed to put a sleep in on the `global` and `globalNodeMesh` checks as they were just being repeatedly called as fast as it could.

That should be it for this, I'll squash if those changes look good to you.

File: backends/k8s/client.go [hunk 10]
```diff
@@ -183,46 +184,49 @@ func (c *Client) WatchPrefix(prefix string, waitIndex uint64, stopChan chan bool
 			ver := tprs.Metadata.ResourceVersion
 			return convertResourceVersionToUint(ver, prefix)
 		default:
-			return 0, nil
+			// We aren't tracking this key, default to 10 second refresh.
+			time.Sleep(10 * time.Second)
```
Comment (caseydavenport) [inline backends/k8s/client.go]: If we're not tracking this key does it make sense to do it even less frequently than 10s?

Comment (heschlie) [inline backends/k8s/client.go]: I think that is sound logic, could just make it 10 minutes? an hour?  It is really just there as more of a safety net than anything.

Comment (caseydavenport) [inline backends/k8s/client.go]: idk, it's kind of a magic number.  

Why don't we add a debug log here and set it to a minute.

File: backends/k8s/client.go [hunk 11]
```diff
@@ -185,7 +185,8 @@ func (c *Client) WatchPrefix(prefix string, waitIndex uint64, stopChan chan bool
 			return convertResourceVersionToUint(ver, prefix)
 		default:
 			// We aren't tracking this key, default to 10 second refresh.
```
Comment (caseydavenport) [inline backends/k8s/client.go]: Comment needs an update now too!

**Comment by heschlie** [comment]
@caseydavenport debug message added, I've also gone ahead and squashed my commits.
